### ISSUE TITLE
Analyzer crashes with unhandled exception when using mixin

### ISSUE THREAD

**Issue by apaatsio** [dart-lang/sdk#34873]
Here's a minimal example I could come up with to reproduce the error:

```dart
import 'dart:async';

mixin MyMixin {}

class MyClass with MyMixin {}

Future<MyClass> func1() => Future.value(func2());

MyClass func2() => null;
 ```

When I run the analyzer I get the following error:

```text
❯ dartanalyzer mixin.dart 
Analyzing mixin.dart...
Unhandled exception:
NoSuchMethodError: The getter 'element' was called on null.
Receiver: null
Tried calling: element
#0      Object.noSuchMethod (dart:core/runtime/libobject_patch.dart:50:5)
#1      GenericInferrer._matchInterfaceSubtypeOf.guardedInterfaceSubtype (package:analyzer/src/generated/type_system.dart:493:29)
#2      GenericInferrer._matchInterfaceSubtypeOf (package:analyzer/src/generated/type_system.dart:511:32)
#3      GenericInferrer._matchInterfaceSubtypeOf.guardedInterfaceSubtype (package:analyzer/src/generated/type_system.dart:494:24)
#4      GenericInferrer._matchInterfaceSubtypeOf (package:analyzer/src/generated/type_system.dart:516:34)
#5      GenericInferrer._matchSubtypeOf (package:analyzer/src/generated/type_system.dart:656:14)
#6      GenericInferrer.tryMatchSubtypeOf (package:analyzer/src/generated/type_system.dart:271:20)
#7      GenericInferrer._matchSubtypeOf (package:analyzer/src/generated/type_system.dart:593:11)
#8      GenericInferrer.tryMatchSubtypeOf (package:analyzer/src/generated/type_system.dart:271:20)
#9      GenericInferrer.constrainArgument (package:analyzer/src/generated/type_system.dart:116:5)
#10     StrongTypeSystemImpl.inferGenericFunctionOrType (package:analyzer/src/generated/type_system.dart:911:16)
...
```

- dartanalyzer version 2.1.0-edge.a50dfd66980ed5c9aef9851fcc57557fb607769c
- Dart VM version: 2.1.0-dev.6.0.flutter-a50dfd6698 (Tue Oct 9 13:47:11 2018 +0000) on "linux_x64"
- OS: Linux

**Comment by bwilkerson** [comment]
@stereotype441 At package:analyzer/src/generated/type_system.dart:511:32, `i1` is most likely a mixin, which, as currently modeled, means that it doesn't have a superclass.

**Comment by devoncarew** [comment]
@stereotype441, can you take this or assign the right owner? It sounds like this is blocking the current Flutter roll.

**Comment by devoncarew** [comment]
@jason-simmons, do you mind attaching the error output from Flutter, and / or linking to a CI log? Thanks!

**Comment by jason-simmons** [comment]
Stack trace from the analyzer:

```
Unhandled exception:
NoSuchMethodError: The getter 'element' was called on null.
Receiver: null
Tried calling: element
#0      Object.noSuchMethod (dart:core/runtime/libobject_patch.dart:50:5)
#1      GenericInferrer._matchInterfaceSubtypeOf.guardedInterfaceSubtype (package:analyzer/src/generated/type_system.dart:493:29)
#2      GenericInferrer._matchInterfaceSubtypeOf (package:analyzer/src/generated/type_system.dart:511:32)
#3      GenericInferrer._matchInterfaceSubtypeOf.guardedInterfaceSubtype (package:analyzer/src/generated/type_system.dart:494:24)
#4      GenericInferrer._matchInterfaceSubtypeOf (package:analyzer/src/generated/type_system.dart:516:34)
#5      GenericInferrer._matchSubtypeOf (package:analyzer/src/generated/type_system.dart:656:14)
#6      GenericInferrer.tryMatchSubtypeOf (package:analyzer/src/generated/type_system.dart:271:20)
#7      GenericInferrer.constrainReturnType (package:analyzer/src/generated/type_system.dart:136:5)
#8      StrongTypeSystemImpl.inferGenericFunctionOrType (package:analyzer/src/generated/type_system.dart:905:16)
#9      ResolverVisitor._inferArgumentTypesForGeneric (package:analyzer/src/generated/resolver.dart:5809:17)
#10     ResolverVisitor._inferArgumentTypesForInvocation (package:analyzer/src/generated/resolver.dart:5884:25)
#11     ResolverVisitor.visitMethodInvocation (package:analyzer/src/generated/resolver.dart:5373:5)
#12     MethodInvocationImpl.accept (package:analyzer/src/dart/ast/ast.dart:7793:49)
#13     NodeListImpl.accept (package:analyzer/src/dart/ast/ast.dart:8290:20)
#14     ArgumentListImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:423:16)
#15     ResolverVisitor.visitNode (package:analyzer/src/generated/resolver.dart:5412:10)
#16     UnifyingAstVisitor.visitArgumentList (package:analyzer/dart/ast/visitor.dart:2945:45)
#17     ResolverVisitor.visitArgumentList (package:analyzer/src/generated/resolver.dart:4517:18)
#18     ArgumentListImpl.accept (package:analyzer/src/dart/ast/ast.dart:419:49)
#19     ResolverVisitor.visitMethodInvocation (package:analyzer/src/generated/resolver.dart:5374:24)
#20     MethodInvocationImpl.accept (package:analyzer/src/dart/ast/ast.dart:7793:49)
#21     ExpressionStatementImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:4232:18)
#22     ResolverVisitor.visitNode (package:analyzer/src/generated/resolver.dart:5412:10)
#23     UnifyingAstVisitor.visitExpressionStatement (package:analyzer/dart/ast/visitor.dart:3052:59)
#24     ExpressionStatementImpl.accept (package:analyzer/src/dart/ast/ast.dart:4228:49)
#25     NodeListImpl.accept (package:analyzer/src/dart/ast/ast.dart:8290:20)
#26     BlockImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:1240:17)
#27     ResolverVisitor.visitNode (package:analyzer/src/generated/resolver.dart:5412:10)
#28     UnifyingAstVisitor.visitBlock (package:analyzer/dart/ast/visitor.dart:2966:31)
#29     ScopedVisitor.visitBlock (package:analyzer/src/generated/resolver.dart:6299:13)
#30     BlockImpl.accept (package:analyzer/src/dart/ast/ast.dart:1236:49)
#31     BlockFunctionBodyImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:1185:13)
#32     ResolverVisitor.visitNode (package:analyzer/src/generated/resolver.dart:5412:10)
#33     UnifyingAstVisitor.visitBlockFunctionBody (package:analyzer/dart/ast/visitor.dart:2969:55)
#34     ScopedVisitor.visitBlockFunctionBody (package:analyzer/src/generated/resolver.dart:6311:13)
#35     ResolverVisitor.visitBlockFunctionBody (package:analyzer/src/generated/resolver.dart:4644:13)
#36     BlockFunctionBodyImpl.accept (package:analyzer/src/dart/ast/ast.dart:1181:49)
#37     MethodDeclarationImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:7681:12)
#38     ResolverVisitor.visitNode (package:analyzer/src/generated/resolver.dart:5412:10)
#39     UnifyingAstVisitor.visitMethodDeclaration (package:analyzer/dart/ast/visitor.dart:3153:55)
#40     ScopedVisitor.visitMethodDeclarationInScope (package:analyzer/src/generated/resolver.dart:6777:11)
#41     ResolverVisitor.visitMethodDeclarationInScope (package:analyzer/src/generated/resolver.dart:5360:11)
#42     ScopedVisitor.visitMethodDeclaration (package:analyzer/src/generated/resolver.dart:6769:7)
#43     ResolverVisitor.visitMethodDeclaration (package:analyzer/src/generated/resolver.dart:5350:13)
#44     MethodDeclarationImpl.accept (package:analyzer/src/dart/ast/ast.dart:7672:49)
#45     NodeListImpl.accept (package:analyzer/src/dart/ast/ast.dart:8290:20)
#46     ScopedVisitor.visitClassMembersInScope (package:analyzer/src/generated/resolver.dart:6380:18)
#47     ScopedVisitor.visitClassDeclaration (package:analyzer/src/generated/resolver.dart:6357:11)
#48     ResolverVisitor.visitClassDeclaration (package:analyzer/src/generated/resolver.dart:4683:13)
#49     ClassDeclarationImpl.accept (package:analyzer/src/dart/ast/ast.dart:1751:49)
#50     ResolverVisitor.visitCompilationUnit (package:analyzer/src/generated/resolver.dart:4747:25)
#51     CompilationUnitImpl.accept (package:analyzer/src/dart/ast/ast.dart:2416:49)
#52     LibraryAnalyzer._resolveFile (package:analyzer/src/dart/analysis/library_analyzer.dart:603:10)
#53     LibraryAnalyzer._analyze.<anonymous closure> (package:analyzer/src/dart/analysis/library_analyzer.dart:107:9)
#54     __InternalLinkedHashMap&_HashVMBase&MapMixin&_LinkedHashMapMixin.forEach (dart:collection/runtime/libcompact_hash.dart:370:8)
#55     LibraryAnalyzer._analyze (package:analyzer/src/dart/analysis/library_analyzer.dart:106:13)
#56     LibraryAnalyzer.analyze.<anonymous closure> (package:analyzer/src/dart/analysis/library_analyzer.dart:83:14)
<asynchronous suspension>
#57     _PerformanceTagImpl.makeCurrentWhileAsync (package:analyzer/src/generated/utilities_general.dart:270:21)
<asynchronous suspension>
#58     LibraryAnalyzer.analyze (package:analyzer/src/dart/analysis/library_analyzer.dart:82:43)
<asynchronous suspension>
#59     AnalysisDriver._computeAnalysisResult.<anonymous closure> (package:analyzer/src/dart/analysis/driver.dart:1248:71)
<asynchronous suspension>
#60     PerformanceLog.runAsync (package:analyzer/src/dart/analysis/performance_logger.dart:52:21)
<asynchronous suspension>
#61     AnalysisDriver._computeAnalysisResult (package:analyzer/src/dart/analysis/driver.dart:1224:20)
<asynchronous suspension>
#62     AnalysisDriver.getErrors (package:analyzer/src/dart/analysis/driver.dart:638:43)
<asynchronous suspension>
#63     BuildMode._printErrors.<anonymous closure> (package:analyzer_cli/src/build_mode.dart:507:43)
<asynchronous suspension>
#64     PerformanceLog.runAsync (package:analyzer/src/dart/analysis/performance_logger.dart:52:21)
<asynchronous suspension>
#65     BuildMode._printErrors (package:analyzer_cli/src/build_mode.dart:495:18)
<asynchronous suspension>
#66     BuildMode.analyze.<anonymous closure> (package:analyzer_cli/src/build_mode.dart:305:15)
<asynchronous suspension>
#67     PerformanceLog.runAsync (package:analyzer/src/dart/analysis/performance_logger.dart:52:21)
<asynchronous suspension>
#68     BuildMode.analyze (package:analyzer_cli/src/build_mode.dart:204:25)
<asynchronous suspension>
#69     Driver._buildModeAnalyze (package:analyzer_cli/src/driver.dart:452:14)
<asynchronous suspension>
#70     Driver.start (package:analyzer_cli/src/driver.dart:194:38)
<asynchronous suspension>
#71     main (dart_lang/v2_1_0_dev_7_1/pkg/analyzer_cli/bin/analyzer.dart:23:17)
<asynchronous suspension>
#72     _startIsolate.<anonymous closure> (dart:isolate/runtime/libisolate_patch.dart:287:32)
#73     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:171:12)
```

**Comment by stereotype441** [comment]
This is a duplicate of #34907, which was fixed on Oct 24 by 0009ca7b1bedf27defbd3fd539a8081fcbb72d40.  @apaatsio's report indicates he was using SDK version a50dfd66980ed5c9aef9851fcc57557fb607769c, which is definitely from before the fix (Oct 9).

@jason-simmons didn't mention which SDK version he was using, but judging by the line numbers in the stack trace it is from before the bug fix (after the fix, package:analyzer/src/generated/type_system.dart:493 no longer contains a reference to `.element`).

Judging by @a-siva's comment, I'm guessing the roll that is blocked is the roll of Flutter into Google internal code; if that's the case then we can fix the problem by either waiting for 2.1.0-dev.8.0 to roll into internal Google code (it has the fix) or by patching in 0009ca7b1bedf27defbd3fd539a8081fcbb72d40 to the internal codebase.  I'll send out a CL via internal channels that patches in 0009ca7b1bedf27defbd3fd539a8081fcbb72d40 to the internal codebase, in case you want to try that.

**Comment by a-siva** [comment]
@jason-simmons was using 2_1_0_dev_7_1 which is based on https://github.com/dart-lang/sdk/commit/a50dfd66980ed5c9aef9851fcc57557fb607769c

@keertip do you know when 2.1.0-dev.8.0 will be available ?

**Comment by keertip** [comment]
Can't release 2.1.0-dev.8.0 till the flutter roll lands. Waiting for fix to [LocalHistoryRoute](https://github.com/flutter/flutter/pull/23382) to land internally.

**Comment by stereotype441** [comment]
0009ca7b1bedf27defbd3fd539a8081fcbb72d40 is now cherry-picked into the internal codebase.  Leaving this bug open until I hear confirmation that the Flutter roll is unblocked.
